corehq/apps/userreports/README.md

```diff
@@ -385,6 +385,76 @@ Variables can be any valid numbers (Python datatypes `int`, `float`, and `long`
 
 More examples can be found on practical [examples page](corehq/apps/userreports/examples/examples.md#evaluator-examples).
 
+#### Case --> Form Lookup
+
+```json
+{
+   "type": "get_case_forms",
+   "case_id_expression": {},  # required
+   "filter": {},  # optional - possibly to remove for now in favor of filter expression
+}
+```
+
+Gets a list of forms from a case, optionally filtering (TBD if that belongs here).
+
+#### Filter Expression
+
+```json
+{
+  "type": "filter",  # not sure on name
+  "items_expression": {},  # expression should return a list
+  "filter_expression": {}  # will be called on each item, should return true or false indicating whether to include
+}
+```
+
+This should work just like python's `filter` function.
+
+#### Map Expression
+
+```json
+{
+  "type": "map",  # not sure on name
+  "items_expression": {},  # expression should return a list
+  "map_expression": {}  # expression to be evaluated on each item in the list
+}
+```
+
+In python the equivalent would be `map(lambda x: x.y, my_list)`, which would return a new list with the lambda function applied.
+
+#### Reduce Expression
+
+```json
+{
+  "type": "reduce",  # not sure on name
+  "items_expression":  {},  # expression should return a list
+  "aggregation_fn": "sum"  # one of a fixed set of allowed aggregation functions
```

Comment:
I see I didn't realize that's what was being proposed in Option 2. I think
making a new first-class sort expression (not building it into map) sounds
like the best plan to me.

On Thu, Mar 31, 2016 at 11:22 AM, Sravan Reddy <notifications@github.com>
wrote:

> In corehq/apps/userreports/README.md
> <https://github.com/dimagi/commcare-hq/pull/10684#discussion_r58023175>:
>
> > +{
> > +  "type": "map",  # not sure on name
> > +  "items_expression": {},  # expression should return a list
> > +  "map_expression": {}  # expression to be evaluated on each item in the list
> > +}
> > +```
> > +
> > +In python the equivalent would be `map(lambda x: x.y, my_list)`, which would return a new list with the lambda function applied.
> > +
> > +#### Reduce Expression
> > +
> > +```json
> > +{
> > +  "type": "reduce",  # not sure on name
> > +  "items_expression":  {},  # expression should return a list
> > +  "aggregation_fn": "sum"  # one of a fixed set of allowed aggregation functions
>
> That was one of the options I added in the doc as well! Another 3rd option
> is a highly custom 'latest-form-property' expression (a shorthand, because
> this expression is the most commonly used)
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/dimagi/commcare-hq/pull/10684/files/e3d996a245afe25960ef31df5ef336c0ee8fa1a5#r58023175>
>